src/test/ui/borrowck/borrowck-assign-comp-idx.stderr

```diff
@@ -21,6 +21,8 @@ LL |         || p[0] = 5);
    |         ^^ - second borrow occurs due to use of `p` in closure
    |         |
    |         mutable borrow occurs here
+   |
+   = help: try adding a local storing the argument and then using the local in the call
```

Comment:
There are many cases like these where the help message is emitted when it doesn't make sense. What heuristics can I use to only emit in the correct situations? (See also my above comment.)

Reply:
Can you categorize the different cases where this is reported wrongly and the cases where it is reported correctly? Maybe we can add another field to BorrowExplanation::UsedLater that carries the information we need

Reply:
These are the categories of false positives I saw where the help was erroneously
emitted, with the more common categories coming first:

- two borrows of the same variable passed to a call (e.g., `foo.method(&mut
  foo)` or `(self.func)(self, arg)`)
- variable passed as argument but variable also mutated in a closure that is
  passed as argument
- closure called with argument that is also mutably borrowed inside the closure

Reply:
These two cases in particular seem tricky to avoid adding the help for:

```
error[E0499]: cannot borrow `*s` as mutable more than once at a time
  --> $DIR/issue-18566.rs:23:19
   |
LL |     MyPtr(s).poke(s);
   |           -  ---- ^ second mutable borrow occurs here
   |           |  |
   |           |  first borrow later used by call
   |           first mutable borrow occurs here
   |
   = help: try adding a local storing the argument and then using the local in the call
```

```
error[E0499]: cannot borrow `thing` as mutable more than once at a time
  --> $DIR/retain-resume-ref.rs:23:25
   |
LL |     gen.as_mut().resume(&mut thing);
   |                         ---------- first mutable borrow occurs here
LL |     gen.as_mut().resume(&mut thing);
   |                  ------ ^^^^^^^^^^ second mutable borrow occurs here
   |                  |
   |                  first borrow later used by call
   |
   = help: try adding a local storing the argument and then using the local in the call
```

Reply:
So... one thing I've noticed going through the examples is that most of the false positives are when the second mutable borrow is an expression that does not contain function calls. So that would be my first step, do not emit the help if there are no function calls.

Additionally the help could get reworded to state that "if the borrows aren't inherently in conflict with each other storing the argument in a local can resolve the situation"

Reply:
> So... one thing I've noticed going through the examples is that most of the false positives are when the second mutable borrow is an expression that does not contain function calls. So that would be my first step, do not emit the help if there are no function calls.

How would I go about doing that? Is there some way to get the (T)HIR version of a MIR statement? Or is there some kind of pre-existing MIR analysis I can call? I haven't worked on the borrow-checker before so I'm new to how you do things like this :)

Reply:
It may possibly suffice to check whether the local being borrowed is a temporary. If it is a temporary that means it was created by an expression and isn't a user-defined local. I think that may be enough, even if a "proper" analysis would have to backtrack through the temporaries' assignments to find function calls.

Reply:
Hmm, now the note is not being emitted when it should. I think the issue is that the place being borrowed is `self`, which is obviously not a temporary.

Here's the patch I applied:

```diff
diff --git a/compiler/rustc_mir/src/borrow_check/diagnostics/conflict_errors.rs b/compiler/rustc_mir/src/borrow_check/diagnostics/conflict_errors.rs
index 8d8d8aa6d6e..4bf3ece7628 100644
--- a/compiler/rustc_mir/src/borrow_check/diagnostics/conflict_errors.rs
+++ b/compiler/rustc_mir/src/borrow_check/diagnostics/conflict_errors.rs
@@ -678,13 +678,13 @@ pub(in crate::borrow_check) fn report_conflicting_borrow(
             None,
         );
 
-        match explanation {
-            BorrowExplanation::UsedLater(LaterUseKind::Call, _call_span) => {
+        if let BorrowExplanation::UsedLater(LaterUseKind::Call, _call_span) = explanation {
+            // Heuristic so this note isn't triggered on user-defined locals.
+            if self.body.local_kind(issued_borrow.borrowed_place.local) == LocalKind::Temp {
                 err.help(
                     "try adding a local storing the argument and then using the local in the call",
                 );
             }
-            _ => {}
         }
 
         err
```

Do you have ideas for other things to try?

Reply:
> Hmm, now the note is not being emitted when it should. I think the issue is that the place being borrowed is `self`, which is obviously not a temporary.

I don't understand, do you have an example that worked before that patch but not after it?

Maybe you need to check for `!= LocalKind::Var`, so that `Arg` is included

Reply:
One thing that may be the problem is that a temporary is created for `&mut some_var`, so in that case it is a temporary, but I would have expected that to cause *more* reports, not less.

Reply:
> I don't understand, do you have an example that worked before that patch but not after it?

```rust
    struct Foo;

    impl Foo {
        fn foo(&mut self, _: f32) -> i32 { todo!() }
        fn bar(&mut self) -> f32 { todo!() }
        fn baz(&mut self) {
            self.foo(self.bar()); //~ ERROR
        }
    }
```

### Before patch

```
error[E0499]: cannot borrow `*self` as mutable more than once at a time
  --> $DIR/suggest-local-var-double-mut.rs:12:22
   |
LL |             self.foo(self.bar());
   |             ---- --- ^^^^ second mutable borrow occurs here
   |             |    |
   |             |    first borrow later used by call
   |             first mutable borrow occurs here
   |
   = help: try adding a local storing the argument and then using the local in the call
```

### After patch

(The note is gone.)

```
error[E0499]: cannot borrow `*self` as mutable more than once at a time
  --> $DIR/suggest-local-var-double-mut.rs:12:22
   |
LL |             self.foo(self.bar());
   |             ---- --- ^^^^ second mutable borrow occurs here
   |             |    |
   |             |    first borrow later used by call
   |             first mutable borrow occurs here
```

Reply:
My approach would be to add tracing statements to the code and see what it says about the variable (and also look at the MIR). I don't understand why this doesn't work tho. The result of the bar call should be stored in a temporary variable... oh... we are checking the wrong thing for temporariness xD. You need to check the argument of the outer function call

Reply:
> oh... we are checking the wrong thing for temporariness xD. You need to check the argument of the outer function call

Oh, that explains it! I think we were checking if `self` is a temporary, which of course it should not be.

How do I get the `Place` of the argument for the outer call? I at first thought I would either use `issued_borrow.assigned_place` or look up the terminator belonging to `issued_borrow.reserve_location` (or maybe `activation_location`?), but based on looking at the generated MIR, I don't think that will give me the right answer. Maybe this needs a full-on MIR analysis?

Here's the MIR for the example I gave earlier (https://github.com/rust-lang/rust/pull/83174#r619567264):

```rust
// MIR for `<impl at foo.rs:3:5: 9:6>::baz` 0 mir_map

fn <impl at foo.rs:3:5: 9:6>::baz(_1: &mut Foo) -> () {
    debug self => _1;                    // in scope 0 at foo.rs:6:16: 6:25
    let mut _0: ();                      // return place in scope 0 at foo.rs:6:27: 6:27
    let _2: i32;                         // in scope 0 at foo.rs:7:13: 7:33
    let mut _3: &mut Foo;                // in scope 0 at foo.rs:7:13: 7:17
    let mut _4: f32;                     // in scope 0 at foo.rs:7:22: 7:32
    let mut _5: &mut Foo;                // in scope 0 at foo.rs:7:22: 7:26

    bb0: {
        StorageLive(_2);                 // scope 0 at foo.rs:7:13: 7:33
        StorageLive(_3);                 // scope 0 at foo.rs:7:13: 7:17
        _3 = &mut (*_1);                 // scope 0 at foo.rs:7:13: 7:17
        StorageLive(_4);                 // scope 0 at foo.rs:7:22: 7:32
        StorageLive(_5);                 // scope 0 at foo.rs:7:22: 7:26
        _5 = &mut (*_1);                 // scope 0 at foo.rs:7:22: 7:26
        _4 = Foo::bar(move _5) -> [return: bb1, unwind: bb3]; // scope 0 at foo.rs:7:22: 7:32
                                         // mir::Constant
                                         // + span: foo.rs:7:27: 7:30
                                         // + literal: Const { ty: for<'r> fn(&'r mut Foo) -> f32 {Foo::bar}, val: Value(Scalar(<ZST>)) }
    }

    bb1: {
        StorageDead(_5);                 // scope 0 at foo.rs:7:31: 7:32
        _2 = Foo::foo(move _3, move _4) -> [return: bb2, unwind: bb3]; // scope 0 at foo.rs:7:13: 7:33
                                         // mir::Constant
                                         // + span: foo.rs:7:18: 7:21
                                         // + literal: Const { ty: for<'r> fn(&'r mut Foo, f32) -> i32 {Foo::foo}, val: Value(Scalar(<ZST>)) }
    }

    bb2: {
        StorageDead(_4);                 // scope 0 at foo.rs:7:32: 7:33
        StorageDead(_3);                 // scope 0 at foo.rs:7:32: 7:33
        StorageDead(_2);                 // scope 0 at foo.rs:7:33: 7:34
        _0 = const ();                   // scope 0 at foo.rs:6:27: 8:10
        return;                          // scope 0 at foo.rs:8:10: 8:10
    }

    bb3 (cleanup): {
        resume;                          // scope 0 at foo.rs:6:9: 8:10
    }
}
```

Reply:
hmm, yea, if _4 does not show up in any of the variables available to you, then you'd need to do something more complicated. I can't really look into this right now though. Probably early July I can dig into it in more depth.

Reply:
@oli-obk Is there a way to look up the statement/terminator where a `Local` is used? I have a potential idea for how we could get this note to be more precise. In the MIR above, I think the borrowck error is reported at this line:

```rust
_5 = &mut (*_1);
```

because `_1` (aka, `self`) is already mutably borrowed. `_5` is then used in `bar` call, and then the result of that call (`_4`) is used in the `foo` call. If we could easily look up the statement where a local is used, then we could theoretically connect all of these statements together to determine that indeed, we should report the note here.

I'm sure this could be done with a new MIR analysis that would be invoked when there was a borrowck error, but it would be nice if we could use the results of a pre-existing analysis. Do you know if those results exist?

Reply:
> Is there a way to look up the statement/terminator where a Local is used?

No, not really. As you already discovered, this needs a new analysis. _1 is being used multiple times, and I'm not sure the that the "previous" use is the right one. Determining what "previous" is is nontrivial if you go across basic blocks. Doing it for just the current block would be as easy as walking backwards in the statement list and visiting the statement to find a specific local. I do know that there is a visitor for finding locals in MIR, so you can re-use that one, but I also don't remember where that visitor is, so you'll have to go through all impls of https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/visit/trait.Visitor.html to find the right one

That said, I believe mir borrowck could track this information, but I wouldn't even know where to start adjusting mir borrowck to know about this and whether it could cause a slowdown.

Reply:
> _1 is being used multiple times, and I'm not sure the that the "previous" use is the right one.

I was actually suggesting that we see where `_5` and `_4` are used; not where `_1` is used. I'm not certain it'll work though :)

> I do know that there is a visitor for finding locals in MIR, so you can re-use that one, but I also don't remember where that visitor is, so you'll have to go through all impls of https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/visit/trait.Visitor.html to find the right one

Sounds good; I'll look for that. Thanks!